### DISCUSSION TITLE
Row level security on views

### DISCUSSION THREAD

**Discussion by programbo** [supabase/supabase#1501] [Questions] [answered]
I couldn't figure out how to use `DISTINCT ON` with the [supabase-js client library](https://supabase.io/docs/guides/client-libraries) but I got around it by creating a view. I have since learned that applying security policies on a view is a problem because it is accessed as the creator so `auth.role() = 'authenticated'` doesn't work. Can anyone suggest a workaround or maybe even an alternative solution for the original `DISTINCT` issue?

**Comment by steve-chavez** [comment, marked as answer]
This is now is supported natively with security invoker views:

```sql
CREATE VIEW sorted_titles WITH (security_invoker) as SELECT title FROM posts ORDER BY title;
```

---

~~If you change the view's owner then RLS will work normally. Like:~~

~~ALTER VIEW <your_view> OWNER TO authenticated;~~

~~For more details, see [here](https://postgrest.org/en/stable/schema_structure.html#views).~~

**Reply by programbo** [reply]
Thanks!

**Reply by chipilov** [reply]
@steve-chavez Just to clarify, the owning role of the view can be ANY role which does NOT bypass RLS, it does NOT need to be the role used in the RLS rule, correct?

That is, in the example above, the view can be handed over to the 'anon' role, and auth.role() == 'authenticated' in the RLS rule would still work as expected because auth.role() is the role from the JWT token and NOT the role with which the DB statement is executed.

**Reply by steve-chavez** [reply]
> Just to clarify, the owning role of the view can be ANY role which does NOT bypass RLS, it does NOT need to be the role used in the RLS rule, correct?

Note that you can specify a role when you create a policy `create policy pol on tbl to <role>`, when role is not specified, the policy is considered as PUBLIC(applied to all roles which don't bypass RLS). So the answer is yes, sorta.

> still work as expected because auth.role() is the role from the JWT token and NOT the role with which the DB statement is executed.

Yes, true about the JWT part. The other part is not true as PostgREST also switches to this role(it does a `set local role to <jwt_role>`). You could check that by doing `select current_user`.

**Reply by chipilov** [reply]
> The other part is not true as PostgREST also switches to this role(it does a set local role to <jwt_role>)

I see. However, if I understand correctly, in the case of views, the JWT role is overridden yet again by PostgreSQL yet again in order to execute it with the role of the view:

1. a request comes in where the JWT role cliam is "authenticated"
2. PostgREST sets the current_user to "authenticated"
3. PostgREST invokes the view from the request
4. PostgreSQL sets the role again to the owner of the view, so the statement is NOT executed with the role of the JWT for the request

Is that correct?

**Reply by HumanBot000** [reply]
> Edit: For PostgreSQL 15, this now is supported natively with security invoker views:
> 
> ```sql
> CREATE VIEW sorted_titles WITH (security_invoker) as SELECT title FROM posts ORDER BY title;
> ```
> 
> If you change the view's owner then RLS will work normally. Like:
> 
> ```sql
> ALTER VIEW <your_view> OWNER TO authenticated;
> ```
> 
> For more details, see [here](https://postgrest.org/en/stable/schema_structure.html#views).



```
GRANT USAGE ON SCHEMA schedule TO authenticated;
GRANT ALL PRIVILEGES ON ALL TABLES IN SCHEMA schedule TO authenticated;

ALTER VIEW schedule.schedule_metadata_with_current_week OWNER TO authenticated;
```

results in ```ERROR:  42501: permission denied for schema schedule```

**Comment by gilbert** [comment]
For those using pgsodium to encrypt a secrets column in a table, you can alter your decrypted view to be SECURITY INVOKER instead of the default SECURITY DEFINER using an alter view statement:

```sql
ALTER VIEW decrypted_table_with_pg_sodium_encryption SET (security_invoker = true);
```

**Reply by rex-ogre** [reply]
Really helpful!!!

**Comment by Magnuti** [comment]
This worked for me

```sql
CREATE OR REPLACE VIEW "public"."view" AS ...

REVOKE ALL ON TABLE "public"."view" FROM "anon";
REVOKE ALL ON TABLE "public"."view" FROM "authenticated";
```

**Reply by steve-chavez** [reply]
JFYI: New projects are coming with a checkbox for disabling default privileges on the public schema for the API roles. With that REVOKEs are not necessary.
